### PR TITLE
Improvements 4.3

### PR DESCRIPTION
**Pull request by lazedo**

### CODE REVIEW

File: kamailio/db_queries_mysql.cfg [hunk 1]
```diff
@@ -11,4 +11,5 @@
 #!substdef "!KZQ_RESET_PUBLISHER_UPDATE!update active_watchers set expires = \$TS where id in (select b.id from presentity a inner join active_watchers b on a.username = b.to_user and a.domain = b.to_domain and a.event = b.event where a.sender = \"\$var(MediaUrl)\")!g"
 #!substdef "!KZQ_PRESENCE_SEARCH_DETAIL!select * from active_watchers_log where presentity_uri = \"\$var(presentity_uri)\"!g"
 #!substdef "!KZQ_PRESENCE_SEARCH_SUMMARY!select * from active_watchers where watcher_domain = \"\$var(Domain)\"!g"
+#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = \"\$subs(to_user)\" and domain = \"\$subs(to_domain)\" and event = \"\$subs(event)\"!g"
```
Comment (sergey-safarov) [inline kamailio/db_queries_mysql.cfg]: Could you add same for postgres macros. Details at https://github.com/2600hz/kazoo-configs-kamailio/pull/88
```
#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = '\$subs(to_user)' and domain = '\$subs(to_domain)' and event = '\$subs(event)'!g"
```

May be order/sort macros?

File: kamailio/db_queries_kazoo.cfg [hunk 2]
```diff
@@ -1,15 +1,16 @@
 ####### Database queries ########
-#!substdef "!KZQ_CHECK_MEDIA_SERVER_INSERT!insert into dispatcher (setid, destination) select \$var(SetId), \"\$var(MediaUrl)\" where not exists(select * from dispatcher where destination = \"\$var(MediaUrl)\")!g"
+#!substdef "!KZQ_CHECK_MEDIA_SERVER_INSERT!insert into dispatcher (setid, destination, flags, attrs, description) select \$var(SetId), \"\$var(MediaUrl)\", \$var(flags), \"\$var(attrs)\", \"added by nodes role\"  where not exists(select * from dispatcher where destination = \"\$var(MediaUrl)\")!g"
```
Comment (sergey-safarov) [inline kamailio/db_queries_kazoo.cfg]: Could you make similar changes for postgres masros?

File: kamailio/db_scripts/db_extra_check.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
+
+do_db_extra_check() {
+
+# location
+KazooDB -db ${DB_CURRENT_DB} "delete from location where socket not like 'udp:%';"
```
Comment (sergey-safarov) [inline kamailio/db_scripts/db_extra_check.sql]: Is it mean TCP and TLS client not more supported?

File: kamailio/db_scripts/db_extra_check.sql [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+KazooDB -db ${DB_CURRENT_DB} "delete from location  where expires > 0 and datetime(expires, 'unixepoch') < datetime('now', '-10 seconds');"
```
Comment (sergey-safarov) [inline kamailio/db_scripts/db_extra_check.sql]: Could you point me to issue that is fixed by force delete regitstration and related info

File: kamailio/db_scripts/db_kazoo-specific [hunk 5]
```diff
@@ -40,6 +41,33 @@ CREATE TABLE active_watchers_log (
     user_agent VARCHAR(255) DEFAULT '' COLLATE NOCASE,
     CONSTRAINT active_watchers_active_watchers_log_idx UNIQUE (presentity_uri, watcher_username, watcher_domain, event)
     );
+    
+    create view presentities as select id, cast(printf("sip:%s@%s",username,domain) as varchar(64)) presentity_uri ,
```
Comment (sergey-safarov) [inline kamailio/db_scripts/db_kazoo-specific]: Is required create same for postgresql by me?

File: kamailio/default.cfg [hunk 6]
```diff
@@ -380,17 +395,20 @@ route[CHECK_RETRANS]
...
 
+
+modparam("htable", "htable", "insane=>size=32;autoexpire=120;initval=0;updateexpire=1;")
+
 route[SANITY_CHECK]
 {
-    ## CVE-2018-14767
-    if($(hdr(To)[1]) != $null) {
-        xlog("second To header not null - dropping message");
+    if($sht(insane=>$si:$sp) > 3) {
```
Comment (sergey-safarov) [inline kamailio/default.cfg]: Several security issues is fixed on latest `kamailio:5.1.6`. `5.1` branch is tested by me.
How about to migrate config to `5.2` branch? Now kamailil/kazoo config is not compatible with `kamailio:5.2` branch.

File: kamailio/default.cfg [hunk 7]
```diff
@@ -824,32 +843,33 @@ failure_route[INTERNAL_FAULT]
...
+    } else if ($T_reply_reason =~ "call barred") {
         xlog("L_INFO", "$ci|log|failure route ignoring call barred\n");
     } else if (isflagset(FLAG_SESSION_PROGRESS)) {
         xlog("L_INFO", "$ci|log|failure route ignoring failure after session progress\n");
-    } else if (t_check_status("403") && $var(reply_reason)=="Forbidden") {
+    } else if (t_check_status("403") && $T_reply_reason=="Forbidden") {
         xlog("L_WARNING", "$ci|log|failure route ignoring. Failed auth from IP $si\n");
     } else if (t_check_status("(401)|(407)|(486)|(403)")) {
-        xlog("L_INFO", "$ci|log|failure route ignoring auth reply $T_reply_code $var(reply_reason)\n");
+        xlog("L_INFO", "$ci|log|failure route ignoring auth reply $T_reply_code $T_reply_reason\n");
```
Comment (sergey-safarov) [inline kamailio/default.cfg]: Could you also look https://github.com/2600hz/kazoo-configs-kamailio/pull/63 PR

File: kamailio/default.cfg [hunk 8]
```diff
@@ -923,11 +965,26 @@ route[ADD_AUTHORIZATION_HEADERS]
...
 
+event_route[evrexec:DEFERRED_INIT]
+{
+    xlog("L_INFO", "processing deferred init\n");
+
+    #!ifdef PRESENCE_ROLE
+    route(PRESENCE_INITIAL_CLEANUP);
+    #!endif
+
+    #!import_file "custom-init.cfg"
```
Comment (sergey-safarov) [inline kamailio/default.cfg]: I cannot locate `custom-init.cfg` in repo.
May be I missed?

File: kamailio/presence-notify.cfg [hunk 9]
```diff
@@ -115,3 +125,29 @@ route[PRESENCE_NOTIFY_AMQP]
     kazoo_publish("omnipresence", "$var(rk)", $var(amqp_payload_request));
     xlog("L_INFO", "$ci|log|sent notify callback for event $subs(event) : $tu\n");
 }
+
+
+route[PRESENCE_CLEANUP]
+{
+    $var(Query) = $_s(DELETE FROM active_watchers WHERE expires > 0 AND datetime(expires, 'unixepoch') < datetime('now', '-90 seconds'););
```
Comment (sergey-safarov) [inline kamailio/presence-notify.cfg]: Why this in not macros files?

File: kamailio/presence-role.cfg [hunk 10]
```diff
@@ -113,6 +130,11 @@ route[HANDLE_SUBSCRIBE]
     exit;
 }
 
+route[DELETE_DUPLICATED_SUBSCRIPTIONS]
+{
+    sql_query("exec", '$_s(DELETE FROM ACTIVE_WATCHERS WHERE FROM_USER = "$subs(from_user)" AND FROM_DOMAIN="$subs(from_domain)" AND EVENT="$subs(event)" AND PRESENTITY_URI="$subs(uri)" AND CALLID <> "$subs(callid)")');
```
Comment (sergey-safarov) [inline kamailio/presence-role.cfg]: May be send to macros files?

File: kamailio/db_scripts/db_extra_check.sql (hunk 4 above)
Comment (sergey-safarov) [reply kamailio/db_scripts/db_extra_check.sql]: This also executed before kamaillio started. Now understand

**Comment by sergey-safarov** [comment]
Is PR ready to test on PostgreSQL database? Or is still developed?

File: kamailio/default.cfg (hunk 6 above)
Comment (lazedo) [reply kamailio/default.cfg]: @sergey-safarov we can't use `5.2` due to changes in `dispatcher` module (still testing)

**Comment by lazedo** [comment]
@sergey-safarov about your comments on the `macros`, these will come in a later pr but not a priority. if you want to take the `endeavor` to get `postgres`/ `mysql` updates asap, we will appreciate that. just wait for the final merge. thanks

File: kamailio/db_scripts/kazoodb-sql.sh [hunk 11]
```diff
@@ -0,0 +1,96 @@
+#!/bin/sh -e
+
```
Comment (sergey-safarov) [inline kamailio/db_scripts/kazoodb-sql.sh]: Why not to add comment about `DB_ENGINE` variable usage? Like this

#####################################################################################
##
## If you want prepare SQL file for PostgreSQL or MySQL server, then need to execute
## DB_ENGINE=postgres ./create-kazoodb-sql.sh
##
#####################################################################################

File: kamailio/db_scripts/db_extra_check.sql (hunk 3 above)
Comment (sergey-safarov) [reply kamailio/db_scripts/db_extra_check.sql]: why not embed this and follow command into kamailio config?
as example define `kamailio:mod-init` (like [htable:mod-init](https://kamailio.org/docs/modules/devel/modules/htable.html#idm1031171732)) and then
```
event_route[kazoo:mod-init] {
    route(STARTUP_CLEAN_LOCATION);
    route(STARTUP_CLEAN_WATCHER);
    ... and etc ...
}
route[STARTUP_CLEAN_LOCATION]
{
    sql_query("exec", 'KZQ_STARTUP_CLEAN_LOCATION_NOT_UDP');
    sql_query("exec", 'KZQ_STARTUP_CLEAN_LOCATION_EXPIRIES');
    sql_query("exec", 'KZQ_STARTUP_CLEAN_LOCATION_ATTRS');
    xlog("L_INFO", "$ci|startup|Cleaned active location table\n");
}
route[STARTUP_CLEAN_WATCHER]
{
    sql_query("exec", 'KZQ_STARTUP_CLEAN_WATCHERS_NOT_UDP');
    sql_query("exec", 'KZQ_STARTUP_CLEAN_WATCHERS_EXPIRIES');
    xlog("L_INFO", "$ci|startup|Cleaned active watchers table\n");
}
```

**Comment by sergey-safarov** [comment]
Looks as interest PR https://github.com/kamailio/kamailio/pull/1782
We can maintain dispatcher table is clean and add FS media servers only at runtime.

**Pull request merged**